File: framework/cei_golang/src/exchanges/binance/binance.go
package binance

import (
    "../../impl"
    "errors"
    "fmt"
)

type CandlestickStream struct {
    Symbol string
}

type RateLimits struct {
    RateLimitType string
    Interval      string
    IntervalNum   int64
    Limit         int64
}

type Symbol struct {
    Symbol             string
    Status             string
    BaseAsset          string
    BaseAssetPrecision int64
    QuoteAsset         string
    QuotePrecision     int64
    OrderTypes         []string
    IcebergAllowed     bool
    OcoAllowed         bool
}

type CandlestickData struct {
    OpenTime                 int64
    Open                     float64
    High                     float64
    Low                      float64
    Close                    float64
    Volume                   float64
    CloseTime                int64
    QuoteAssetVolume         float64
    NumberOfTrades           int64
    TakerBuyBaseAssetVolume  float64
    TakerBuyQuoteAssetVolume float64
}

type CandlestickDataList struct {
    Data []CandlestickData
}

type AggregateTrade struct {
    Id           int64
    Price        float64
    Qty          float64
    Timestamp    int64
    FirstTradeId int64
    LastTradeId  int64
    IsBuyerMaker bool
    IsBestMatch  bool
}

type ExchangeInfo struct {
    Timezone   string
    ServerTime int64
    RateLimits []RateLimits
    Symbols    []Symbol
}

type Timestamp struct {
    ServerTime int64
}

type Trade struct {
    Id           int64
    Price        float64
    Qty          float64
    QuoteQty     float64
    Time         int64
    IsBuyerMaker bool
    IsBestMatch  bool
}

type TradeList struct {
    Data []Trade
}

type DepthEntity struct {
    Price float64
    Qty   float64
}

type Depth struct {
    LastUpdateId int64
    Bids         []DepthEntity
    Asks         []DepthEntity
}

type AggregateTradeList struct {
    Data []AggregateTrade
}

type MarketClient struct {
    option *impl.RestfulOptions
}

func NewMarketClient(option *impl.RestfulOptions) *MarketClient {
    inst := new(MarketClient)
    if option != nil {
        inst.option = option
    } else {
        inst.option.Url = "https://api.binance.com"
    }
    return inst
}

func (inst *MarketClient) GetTimestamp() (data Timestamp , exception error) {
    defer func() {
        if err := recover(); err != nil {
            exception = errors.New(err.(string))
        }
    }()
    request := impl.NewRestfulRequest(inst.option)
    request.SetTarget("/api/v3/time")
    request.SetMethod(impl.GET)
    response := impl.RestfulQuery(request)
    rootObj := impl.ParseJsonFromString(response.GetString())
    timestampVar := Timestamp{}
    timestampVar.ServerTime = rootObj.GetInt64("serverTime")
    return timestampVar, nil
}

func (inst *MarketClient) GetAggregateTrades(symbol string, fromId int64, startTime int64, endTime int64, limit int64) (data AggregateTradeList , exception error) {
    defer func() {
        if err := recover(); err != nil {
            exception = errors.New(err.(string))
        }
    }()
    request := impl.NewRestfulRequest(inst.option)
    request.SetTarget("/api/v3/aggTrades")
    request.SetMethod(impl.GET)
    request.AddQueryString("symbol", symbol)
    request.AddQueryString("fromId", impl.ToString(fromId))
    request.AddQueryString("startTime", impl.ToString(startTime))
    request.AddQueryString("startTime", impl.ToString(endTime))
    request.AddQueryString("limit", impl.ToString(limit))
    response := impl.RestfulQuery(request)
    rootObj := impl.ParseJsonFromString(response.GetString())
    aggregateTradeListVar := AggregateTradeList{}
    for _, item := range rootObj.Array() {
        aggregateTradeVar := AggregateTrade{}
        aggregateTradeVar.Id = item.GetInt64("a")
        aggregateTradeVar.Price = item.GetFloat64("p")
        aggregateTradeVar.Qty = item.GetFloat64("q")
        aggregateTradeVar.FirstTradeId = item.GetInt64("f")
        aggregateTradeVar.LastTradeId = item.GetInt64("l")
        aggregateTradeVar.Timestamp = item.GetInt64("T")
        aggregateTradeVar.IsBuyerMaker = item.GetBool("m")
        aggregateTradeVar.IsBestMatch = item.GetBool("M")
        aggregateTradeListVar.Data = append(aggregateTradeListVar.Data, aggregateTradeVar)
    }
    return aggregateTradeListVar, nil
}

func (inst *MarketClient) GetTrade(symbol string, limit int64) (data TradeList , exception error) {
    defer func() {
        if err := recover(); err != nil {
            exception = errors.New(err.(string))
        }
    }()
    request := impl.NewRestfulRequest(inst.option)
    request.SetTarget("/api/v3/trades")
    request.SetMethod(impl.GET)
    request.AddQueryString("symbol", symbol)
    request.AddQueryString("limit", impl.ToString(limit))
    response := impl.RestfulQuery(request)
    rootObj := impl.ParseJsonFromString(response.GetString())
    tradeListVar := TradeList{}
    for _, item := range rootObj.Array() {
        tradeVar := Trade{}
        tradeVar.Id = item.GetInt64("id")
        tradeVar.Price = item.GetFloat64("price")
        tradeVar.Qty = item.GetFloat64("qty")
        tradeVar.QuoteQty = item.GetFloat64("quoteQty")
        tradeVar.Time = item.GetInt64("time")
        tradeVar.IsBuyerMaker = item.GetBool("isBuyerMaker")
        tradeVar.IsBestMatch = item.GetBool("isBestMatch")
        tradeListVar.Data = append(tradeListVar.Data, tradeVar)
    }
    return tradeListVar, nil
}

func (inst *MarketClient) HistoricalTrades(symbol string, limit int64, fromId int64) (data TradeList , exception error) {
    defer func() {
        if err := recover(); err != nil {
            exception = errors.New(err.(string))
        }
    }()
    request := impl.NewRestfulRequest(inst.option)
    request.SetTarget("/api/v3/historicalTrades")
    request.SetMethod(impl.GET)
    request.AddQueryString("symbol", symbol)
    request.AddQueryString("limit", impl.ToString(limit))
    request.AddQueryString("fromId", impl.ToString(fromId))
    response := impl.RestfulQuery(request)
    rootObj := impl.ParseJsonFromString(response.GetString())
    tradeListVar := TradeList{}
    for _, item := range rootObj.Array() {
        tradeVar := Trade{}
        tradeVar.Id = item.GetInt64("id")
        tradeVar.Price = item.GetFloat64("price")
        tradeVar.Qty = item.GetFloat64("qty")
        tradeVar.QuoteQty = item.GetFloat64("quoteQty")
        tradeVar.Time = item.GetInt64("time")
        tradeVar.IsBuyerMaker = item.GetBool("isBuyerMaker")
        tradeVar.IsBestMatch = item.GetBool("isBestMatch")
        tradeListVar.Data = append(tradeListVar.Data, tradeVar)
    }
    return tradeListVar, nil
}

func (inst *MarketClient) GetExchangeInfo() (data ExchangeInfo , exception error) {
    defer func() {
        if err := recover(); err != nil {
            exception = errors.New(err.(string))
        }
    }()
    request := impl.NewRestfulRequest(inst.option)
    request.SetTarget("/api/v3/exchangeInfo")
    request.SetMethod(impl.GET)
    response := impl.RestfulQuery(request)
    rootObj := impl.ParseJsonFromString(response.GetString())
    exchangeInfoVar := ExchangeInfo{}
    exchangeInfoVar.Timezone = rootObj.GetString("timezone")
    exchangeInfoVar.ServerTime = rootObj.GetInt64("serverTime")
    obj := rootObj.GetArray("rateLimits")
    for _, item := range obj.Array() {
        rateLimitsVar := RateLimits{}
        rateLimitsVar.RateLimitType = item.GetString("rateLimitType")
        rateLimitsVar.Interval = item.GetString("interval")
        rateLimitsVar.IntervalNum = item.GetInt64("intervalNum")
        rateLimitsVar.Limit = item.GetInt64("limit")
        exchangeInfoVar.RateLimits = append(exchangeInfoVar.RateLimits, rateLimitsVar)
    }
    obj0 := rootObj.GetArray("symbols")
    for _, item1 := range obj0.Array() {
        symbolVar := Symbol{}
        symbolVar.Symbol = item1.GetString("symbol")
        symbolVar.Status = item1.GetString("status")
        symbolVar.BaseAsset = item1.GetString("baseAsset")
        symbolVar.BaseAssetPrecision = item1.GetInt64("baseAssetPrecision")
        symbolVar.QuoteAsset = item1.GetString("quoteAsset")
        symbolVar.QuotePrecision = item1.GetInt64("quotePrecision")
        symbolVar.OrderTypes = item1.GetStringArray("orderTypes")
        symbolVar.IcebergAllowed = item1.GetBool("icebergAllowed")
        symbolVar.OcoAllowed = item1.GetBool("ocoAllowed")
        exchangeInfoVar.Symbols = append(exchangeInfoVar.Symbols, symbolVar)
    }
    return exchangeInfoVar, nil
}

func (inst *MarketClient) GetDepth(symbol string, limit int64) (data Depth , exception error) {
    defer func() {
        if err := recover(); err != nil {
            exception = errors.New(err.(string))
        }
    }()
    request := impl.NewRestfulRequest(inst.option)
    request.SetTarget("/api/v3/depth")
    request.SetMethod(impl.GET)
    request.AddQueryString("symbol", symbol)
    request.AddQueryString("limit", impl.ToString(limit))
    response := impl.RestfulQuery(request)
    rootObj := impl.ParseJsonFromString(response.GetString())
    depthVar := Depth{}
    depthVar.LastUpdateId = rootObj.GetInt64("lastUpdateId")
    obj := rootObj.GetArray("bids")
    for _, item := range obj.Array() {
        depthEntityVar := DepthEntity{}
        depthEntityVar.Price = item.GetFloat64("[0]")
        depthEntityVar.Qty = item.GetFloat64("[1]")
        depthVar.Bids = append(depthVar.Bids, depthEntityVar)
    }
    obj0 := rootObj.GetArray("asks")
    for _, item1 := range obj0.Array() {
        depthEntityVar2 := DepthEntity{}
        depthEntityVar2.Price = item1.GetFloat64("[0]")
        depthEntityVar2.Qty = item1.GetFloat64("[1]")
        depthVar.Asks = append(depthVar.Asks, depthEntityVar2)
    }
    return depthVar, nil
}

func (inst *MarketClient) GetCandlestickData(symbol string, interval string, startTime int64, endTime int64, limit int64) (data CandlestickDataList , exception error) {
    defer func() {
        if err := recover(); err != nil {
            exception = errors.New(err.(string))
        }
    }()
    request := impl.NewRestfulRequest(inst.option)
    request.SetTarget("/api/v3/klines")
    request.SetMethod(impl.GET)
    request.AddQueryString("symbol", symbol)
    request.AddQueryString("interval", interval)
    request.AddQueryString("startTime", impl.ToString(startTime))
    request.AddQueryString("endTime", impl.ToString(endTime))
    request.AddQueryString("limit", impl.ToString(limit))
    response := impl.RestfulQuery(request)
    rootObj := impl.ParseJsonFromString(response.GetString())
    candlestickDataListVar := CandlestickDataList{}
    for _, item := range rootObj.Array() {
        candlestickDataVar := CandlestickData{}
        candlestickDataVar.Open = item.GetFloat64("\\[0]")
        candlestickDataVar.High = item.GetFloat64("\\[1]")
        candlestickDataListVar.Data = append(candlestickDataListVar.Data, candlestickDataVar)
    }
    return candlestickDataListVar, nil
}


type CandlestickChannel struct {
    option     *impl.WebSocketOptions
    connection *impl.WebSocketConnection
}

func NewCandlestickChannel(option *impl.WebSocketOptions) *CandlestickChannel {
    inst := new(CandlestickChannel)
    if option != nil {
        inst.option = option
    } else {
        inst.option.Url = "wss://stream.binance.com:9443"
    }
    return inst
}

func (inst *CandlestickChannel) Open(symbolList []string, interval string) {
    streams := impl.StringWrapper{}
    streams.AddStringArray(symbolList, true)
    streams.CombineStringItems("", fmt.Sprintf("@kline_%s", interval), "/")
    inst.connection.Connect(fmt.Sprintf("/ws/%s", streams.ToString()))
}

func (inst *CandlestickChannel) SetOnCandlestickHandler(onCandlestick func (data CandlestickStream)) {
    onCandlestickEvent := impl.NewWebSocketEvent(true)
    onCandlestickEvent.SetTrigger(func(msg *impl.WebSocketMessage) bool {
        rootObj := impl.ParseJsonFromString(msg.GetString())
        jsonChecker := new(impl.JsonChecker)
        jsonChecker.CheckEqual("e", "kline", rootObj)
        return jsonChecker.Complete()
    })
    onCandlestickEvent.SetEvent(func(connection *impl.WebSocketConnection, msg *impl.WebSocketMessage)  {
        rootObj := impl.ParseJsonFromString(msg.GetString())
        candlestickStreamVar := CandlestickStream{}
        candlestickStreamVar.Symbol = rootObj.GetString("s")
        onCandlestick(candlestickStreamVar)
    })
    inst.connection.RegisterEvent(onCandlestickEvent)
}


func restfulAuth(request *impl.RestfulRequest, option *impl.RestfulOptions) {
    request.AddHeaderString("X-MBX-APIKEY", option.ApiKey)
    ts := impl.GetNow("Unix_ms")
    request.AddQueryString("timestamp", ts)
    queryString := impl.CombineQueryString(request, impl.NONE, "&")
    postBody := impl.GetRequestInfo(request, impl.POSTBODY, impl.NONE)
    buffer := impl.StringWrapper{}
    buffer.AppendStringItem(queryString)
    buffer.AppendStringItem(postBody)
    buffer.CombineStringItems("", "", "")
    hmac := impl.HMACSHA256(buffer.ToString(), option.SecretKey)
    output := impl.EncodeHex(hmac)
    request.AddQueryString("signature", output)
}

File: framework/cei_cpp/src/exchanges/binance.cpp
#include "exchanges/binance.hpp"
namespace binance {
    Timestamp MarketClient::getTimestamp() {
        RestfulRequest request{this->option};
        request.setTarget("/api/v3/time");
        request.setMethod(RestfulRequest.Method.GET);
        RestfulResponse response = request.query();
        JsonWrapper rootObj = JsonWrapper.parseFromString(response.getString());
        Timestamp timestampVar{};
        timestampVar.serverTime = rootObj.getLong("serverTime");
    }
    AggregateTradeList MarketClient::getAggregateTrades(const std::string& symbol, const long& fromId, const long& startTime, const long& endTime, const long& limit) {
        RestfulRequest request{this->option};
        request.setTarget("/api/v3/aggTrades");
        request.setMethod(RestfulRequest.Method.GET);
        request.addQueryString("symbol", symbol);
        request.addQueryString("fromId", Long.toString(fromId));
        request.addQueryString("startTime", Long.toString(startTime));
        request.addQueryString("startTime", Long.toString(endTime));
        request.addQueryString("limit", Long.toString(limit));
        RestfulResponse response = request.query();
        JsonWrapper rootObj = JsonWrapper.parseFromString(response.getString());
        AggregateTradeList aggregateTradeListVar{};
        for (item : rootObj.Array()) {
            AggregateTrade aggregateTradeVar{};
            aggregateTradeVar.id = item.getLong("a");
            aggregateTradeVar.price = item.getDecimal("p");
            aggregateTradeVar.qty = item.getDecimal("q");
            aggregateTradeVar.firstTradeId = item.getLong("f");
            aggregateTradeVar.lastTradeId = item.getLong("l");
            aggregateTradeVar.timestamp = item.getLong("T");
            aggregateTradeVar.isBuyerMaker = item.getBoolean("m");
            aggregateTradeVar.isBestMatch = item.getBoolean("M");
            aggregateTradeListVar.data = aggregateTradeListVar.data.push_back(aggregateTradeVar);
        }
    }
    TradeList MarketClient::getTrade(const std::string& symbol, const long& limit) {
        RestfulRequest request{this->option};
        request.setTarget("/api/v3/trades");
        request.setMethod(RestfulRequest.Method.GET);
        request.addQueryString("symbol", symbol);
        request.addQueryString("limit", Long.toString(limit));
        RestfulResponse response = request.query();
        JsonWrapper rootObj = JsonWrapper.parseFromString(response.getString());
        TradeList tradeListVar{};
        for (item : rootObj.Array()) {
            Trade tradeVar{};
            tradeVar.id = item.getLong("id");
            tradeVar.price = item.getDecimal("price");
            tradeVar.qty = item.getDecimal("qty");
            tradeVar.quoteQty = item.getDecimal("quoteQty");
            tradeVar.time = item.getLong("time");
            tradeVar.isBuyerMaker = item.getBoolean("isBuyerMaker");
            tradeVar.isBestMatch = item.getBoolean("isBestMatch");
            tradeListVar.data = tradeListVar.data.push_back(tradeVar);
        }
    }
    TradeList MarketClient::historicalTrades(const std::string& symbol, const long& limit, const long& fromId) {
        RestfulRequest request{this->option};
        request.setTarget("/api/v3/historicalTrades");
        request.setMethod(RestfulRequest.Method.GET);
        request.addQueryString("symbol", symbol);
        request.addQueryString("limit", Long.toString(limit));
        request.addQueryString("fromId", Long.toString(fromId));
        RestfulResponse response = request.query();
        JsonWrapper rootObj = JsonWrapper.parseFromString(response.getString());
        TradeList tradeListVar{};
        for (item : rootObj.Array()) {
            Trade tradeVar{};
            tradeVar.id = item.getLong("id");
            tradeVar.price = item.getDecimal("price");
            tradeVar.qty = item.getDecimal("qty");
            tradeVar.quoteQty = item.getDecimal("quoteQty");
            tradeVar.time = item.getLong("time");
            tradeVar.isBuyerMaker = item.getBoolean("isBuyerMaker");
            tradeVar.isBestMatch = item.getBoolean("isBestMatch");
            tradeListVar.data = tradeListVar.data.push_back(tradeVar);
        }
    }
    ExchangeInfo MarketClient::getExchangeInfo() {
        RestfulRequest request{this->option};
        request.setTarget("/api/v3/exchangeInfo");
        request.setMethod(RestfulRequest.Method.GET);
        RestfulResponse response = request.query();
        JsonWrapper rootObj = JsonWrapper.parseFromString(response.getString());
        ExchangeInfo exchangeInfoVar{};
        exchangeInfoVar.timezone = rootObj.getString("timezone");
        exchangeInfoVar.serverTime = rootObj.getLong("serverTime");
        JsonWrapper obj = rootObj.getArray("rateLimits");
        for (item : obj.Array()) {
            RateLimits rateLimitsVar{};
            rateLimitsVar.rateLimitType = item.getString("rateLimitType");
            rateLimitsVar.interval = item.getString("interval");
            rateLimitsVar.intervalNum = item.getLong("intervalNum");
            rateLimitsVar.limit = item.getLong("limit");
            exchangeInfoVar.rateLimits = exchangeInfoVar.rateLimits.push_back(rateLimitsVar);
        }
        JsonWrapper obj0 = rootObj.getArray("symbols");
        for (item1 : obj0.Array()) {
            Symbol symbolVar{};
            symbolVar.symbol = item1.getString("symbol");
            symbolVar.status = item1.getString("status");
            symbolVar.baseAsset = item1.getString("baseAsset");
            symbolVar.baseAssetPrecision = item1.getLong("baseAssetPrecision");
            symbolVar.quoteAsset = item1.getString("quoteAsset");
            symbolVar.quotePrecision = item1.getLong("quotePrecision");
            symbolVar.orderTypes = item1.getStringArray("orderTypes");
            symbolVar.icebergAllowed = item1.getBoolean("icebergAllowed");
            symbolVar.ocoAllowed = item1.getBoolean("ocoAllowed");
            exchangeInfoVar.symbols = exchangeInfoVar.symbols.push_back(symbolVar);
        }
    }
    Depth MarketClient::getDepth(const std::string& symbol, const long& limit) {
        RestfulRequest request{this->option};
        request.setTarget("/api/v3/depth");
        request.setMethod(RestfulRequest.Method.GET);
        request.addQueryString("symbol", symbol);
        request.addQueryString("limit", Long.toString(limit));
        RestfulResponse response = request.query();
        JsonWrapper rootObj = JsonWrapper.parseFromString(response.getString());
        Depth depthVar{};
        depthVar.lastUpdateId = rootObj.getLong("lastUpdateId");
        JsonWrapper obj = rootObj.getArray("bids");
        for (item : obj.Array()) {
            DepthEntity depthEntityVar{};
            depthEntityVar.price = item.getDecimal("[0]");
            depthEntityVar.qty = item.getDecimal("[1]");
            depthVar.bids = depthVar.bids.push_back(depthEntityVar);
        }
        JsonWrapper obj0 = rootObj.getArray("asks");
        for (item1 : obj0.Array()) {
            DepthEntity depthEntityVar2{};
            depthEntityVar2.price = item1.getDecimal("[0]");
            depthEntityVar2.qty = item1.getDecimal("[1]");
            depthVar.asks = depthVar.asks.push_back(depthEntityVar2);
        }
    }
    CandlestickDataList MarketClient::getCandlestickData(const std::string& symbol, const std::string& interval, const long& startTime, const long& endTime, const long& limit) {
        RestfulRequest request{this->option};
        request.setTarget("/api/v3/klines");
        request.setMethod(RestfulRequest.Method.GET);
        request.addQueryString("symbol", symbol);
        request.addQueryString("interval", interval);
        request.addQueryString("startTime", Long.toString(startTime));
        request.addQueryString("endTime", Long.toString(endTime));
        request.addQueryString("limit", Long.toString(limit));
        RestfulResponse response = request.query();
        JsonWrapper rootObj = JsonWrapper.parseFromString(response.getString());
        CandlestickDataList candlestickDataListVar{};
        for (item : rootObj.Array()) {
            CandlestickData candlestickDataVar{};
            candlestickDataVar.open = item.getDecimal("\\[0]");
            candlestickDataVar.high = item.getDecimal("\\[1]");
            candlestickDataListVar.data = candlestickDataListVar.data.push_back(candlestickDataVar);
        }
    }


}
}
